docs/spec/tuples.rst

```diff
@@ -0,0 +1,155 @@
+    t2: tuple[*tuple[str, *tuple[str, ...]]]  # OK
+    t3: tuple[*tuple[str, ...], *tuple[int, ...]]  # Type error
+    t4: tuple[*tuple[str, *tuple[str, ...]], *tuple[int, ...]]  # Type error
+
+An unpacked TypeVarTuple counts as an unbounded tuple in the context of this rule::
+
+    def func[*Ts](t: tuple[*Ts]):
+        t5: tuple[*tuple[str], *Ts]  # OK
+        t6: tuple[*tuple[str, ...], *Ts]  # Type error
+
+The ``*`` syntax requires Python 3.11 or newer. For older versions of Python,
+the ``typing.Unpack`` special form can be used:
+``tuple[int, Unpack[tuple[str, ...]], int]``.
+
+Unpacked tuples can also be used for ``*args`` parameters in a function
+signature: ``def f(*args: *tuple[int, str]): ...``. Unpacked tuples
+can also be used for specializing generic classes or type variables that are
+parameterized using a ``TypeVarTuple``. For more details, see
+:ref:`args_as_typevartuple`.
+
+
+Type Compatibility Rules
+------------------------
+
+Because tuple contents are immutable, the element types of a tuple are covariant.
+For example, ``tuple[int, int]`` is a subtype of ``tuple[float, complex]``.
+
+As discussed above, a homogeneous tuple of arbitrary length is equivalent
+to a union of tuples of different lengths. That means ``tuple[()]``, 
+``tuple[int]`` and ``tuple[int, *tuple[int, ...]]`` are all subtypes of
+``tuple[int, ...]``. The converse is not true; ``tuple[int, ...]``` is not a
+subtype of ``tuple[int]``.
+
+The type ``tuple[Any, ...]`` is bidirectionally compatible with any tuple::
+
+    def func(t1: tuple[int], t2: tuple[int, ...], t3: tuple[Any, ...]):
+        v1: tuple[int, ...] = t1  # OK
+        v2: tuple[Any, ...] = t1  # OK
+
+        v3: tuple[int] = t2  # Type error
+        v4: tuple[Any, ...] = t2  # OK
+
+        v5: tuple[float, float] = t3  # OK
+        v6: tuple[int, *tuple[str, ...]] = t3  # OK
+
+
+The length of a tuple at runtime is immutable, so it is safe for type checkers
+to use length checks to narrow the type of a tuple::
+
+    def func(val: tuple[int] | tuple[str, str] | tuple[int, *tuple[str], int]):
+        if len(val) == 1:
+            # Type can be narrowed to tuple[int].
+            reveal_type(val)  # tuple[int]
+        
+        if len(val) == 2:
+            # Type can be narrowed to tuple[str, str] | tuple[int, int].
+            reveal_type(val)  # tuple[str, str] | tuple[int, int]
+
+        if len(val) == 3:
+            # Type can be narrowed to tuple[int, str, int].
+            reveal_type(val)  # tuple[int, str, int]
+
+This property may also be used to safely narrow tuple types within a ``match``
+statement that uses sequence patterns.
+
+If a tuple element is a union type, the tuple can be safely expanded into a
+union of tuples. For example, ``tuple[int | str]`` is equivalent to 
+``tuple[int] | tuple[str]``. If multiple elements are union types, the expansion
+must consider all combinations. For example, ``tuple[int | str, int | str]`` is
```

Comment:
"must" is a little strong here for behaviour that is exponential

Reply:
I'll change the text to read "...full expansion must consider...". You're correct that partial expansion wouldn't require this. Pyright's current implementation performs partial expansion only (i.e. a single tuple element is expanded rather than multiple), and that's how I avoid the combinatorial explosion.